forc-plugins/forc-node/src/chain_config.rs

```diff
@@ -0,0 +1,574 @@
+pub enum ChainConfig {
+    Local,
+    Testnet,
+    Ignition,
+}
+
+impl Display for ChainConfig {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ChainConfig::Local => write!(f, "local"),
+            ChainConfig::Testnet => write!(f, "testnet"),
+            ChainConfig::Ignition => write!(f, "ignition"),
+        }
+    }
+}
+
+impl From<ChainConfig> for PathBuf {
+    fn from(value: ChainConfig) -> Self {
+        let folder_name = match value {
+            ChainConfig::Local => LOCAL_CONFIG_FOLDER_NAME,
+            ChainConfig::Testnet => TESTNET_CONFIG_FOLDER_NAME,
+            ChainConfig::Ignition => IGNITION_CONFIG_FOLDER_NAME,
+        };
+        user_forc_directory().join(CONFIG_FOLDER).join(folder_name)
+    }
+}
+
+/// A github api, content query response.
+/// Mainly used for fetching a download url and hash for configuration files.
+#[derive(Serialize, Deserialize, Debug)]
+struct GithubContentDetails {
+    name: String,
+    sha: String,
+    download_url: Option<String>,
+    #[serde(rename = "type")]
+    content_type: String,
+}
+
+/// `ConfigFetcher` is responsible for github api integration related to the
+/// configuration operations.
+/// Basically checks remote hash of the corresponding chain configuration.
+/// If there is a mismatch between local and remote instance, overrides the
+/// local instance with remote changes for testnet and mainnet configurations.
+///
+/// For local chain configuration, we only check for existence of it locally.
+/// If the local chain configuration is missing in user's local,
+/// `ConfigFetcher` fetches it but remote updates are not tracked for it.
+pub struct ConfigFetcher {
+    client: reqwest::Client,
+    #[cfg(test)]
+    base_url: String,
+    config_vault: PathBuf,
+}
+
+impl ConfigFetcher {
+    /// Creates a new fetcher to interact with github.
+    /// By default user's chain configuration vault is at: `~/.forc/chainspecs`
+    pub fn new() -> Self {
+        Self {
+            client: reqwest::Client::new(),
+            #[cfg(test)]
+            base_url: "https://api.github.com".to_string(),
+            config_vault: user_forc_directory().join(CONFIG_FOLDER),
+        }
+    }
+
+    #[cfg(test)]
+    /// Override the base url, to be used in tests.
+    pub fn with_base_url(base_url: String) -> Self {
+        Self {
+            client: reqwest::Client::new(),
+            base_url,
+            config_vault: user_forc_directory().join(CONFIG_FOLDER),
+        }
+    }
+
+    #[cfg(test)]
+    /// To OVER
+    pub fn with_test_config(base_url: String, config_vault: PathBuf) -> Self {
+        Self {
+            client: reqwest::Client::new(),
+            base_url,
+            config_vault,
+        }
+    }
+
+    fn get_base_url(&self) -> &str {
+        #[cfg(not(test))]
+        return "https://api.github.com";
+
+        #[cfg(test)]
+        return &self.base_url;
```

Comment:
not sure if we need feature gate here since `base_url` is set in `new` for test env. 
```suggestion
        return &self.base_url;
```

Reply:
Nice one!, cleaned this